Duplicati/Library/Snapshots/USNJournal.cs

```diff
@@ -1,650 +1,652 @@
+
+                m_volumeRootRefNumber = GetFileRefNumber(volumeRoot);
+            }
+            catch
+            {
+                Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Gets the USN JournalID for the current volume
+        /// </summary>
+        public long JournalId => m_journal.UsnJournalID;
+
+        /// <summary>
+        /// Gets the next USN number for the volume
+        /// </summary>
+        public long NextUsn => m_journal.NextUsn;
+
+        /// <summary>
+        /// Returns a list of files or folders that have changed since the recorded USN
+        /// </summary>
+        /// <param name="sourceFileOrFolder">The file or folder to find entries for</param>
+        /// <param name="minUsn">Minimum USN of entry</param>
+        /// <returns>A list of tuples with changed files and folders and their type</returns>
+        public IEnumerable<Tuple<string, EntryType>> GetChangedFileSystemEntries(string sourceFileOrFolder, long minUsn)
+        {
+            return GetChangedFileSystemEntries(sourceFileOrFolder, minUsn, ChangeReason.Any);
+        }
+
+        /// <summary>
+        /// Returns a list of files or folders that have changed since the recorded USN
+        /// </summary>
+        /// <param name="sourceFileOrFolder">The file or folder to find entries for</param>
+        /// <param name="minUsn">Minimum USN of entry</param>
+        /// <param name="reason">Filter expression for change reason</param>
+        /// <returns>A list of tuples with changed files and folders and their type</returns>
+        public IEnumerable<Tuple<string, EntryType>> GetChangedFileSystemEntries(string sourceFileOrFolder, long minUsn, ChangeReason reason)
+        {
+            var isFolder = sourceFileOrFolder.EndsWith(Utility.Utility.DirectorySeparatorString, StringComparison.Ordinal);
+
+            foreach (var r in GetRecords(minUsn))
+            {
+                if (r.UsnRecord.Usn >= minUsn
+                    && (reason == ChangeReason.Any || (MapChangeReason(r.UsnRecord.Reason) & reason) != 0)
+                    && (r.FullPath.Equals(sourceFileOrFolder, Utility.Utility.ClientFilenameStringComparison)
+                        || isFolder && Utility.Utility.IsPathBelowFolder(r.FullPath, sourceFileOrFolder)))
+                {
+                    yield return Tuple.Create(r.FullPath,
+                        r.UsnRecord.FileAttributes.HasFlag(Win32USN.FileAttributes.Directory)
+                            ? EntryType.Directory
+                            : EntryType.File);
+                }
+            }
+        }
+
+        public static string GetVolumeRootFromPath(string path)
+        {
+            if (path == null)
+                throw new Exception(Strings.USNHelper.UnexpectedPathFormat);
+
+            return AlphaFS.Path.GetPathRoot(path);

```

Comment:
In other places, I use a custom version of `System.IO`, like this:

```csharp
private static SystemIOWindows IO_WIN = new SystemIOWindows();
```

It does not yet have `GetPathRoot()` but it has many of the other methods, and we can add `GetPathRoot()` to the `ISystemIO` interface.

Reply:
Agreed, better to decouple this.